### PR TITLE
Chapter 3 Spectral Clustering

### PR DESCRIPTION
**Pull request by tzm41**
Sorry for the delay. It has also been quite hard for me to do this alone, when I don't have a teammate to discuss with. Frankly, the convergence paper is a bit hard for me to comprehend. Maybe my brain is already too fried to take in new theorem proofs.

Please let give me hints on how to parse Theorem 4 and 5 in that paper if possible!

### CODE REVIEW

**Comment by all2187** [comment]
Hi Colin. No worries, I understand (and the extraordinary circumstances around the virus can't have helped). I'll have a look in the next few days and let you know what I think / how to understand the theorems. 

Thank you for your hard work!

File: chapter_3/2_spectral_clustering.tex [hunk 1]
```diff
@@ -153,13 +154,242 @@ \section{Spectral Clustering}
...
+		\begin{split}
+			T_{n}: L_{2}\left(P_{n}\right) \rightarrow L_{2}\left(P_{n}\right), T_{n} f(x)=\int h(x, y) f(y) d P_{n}(y) \\
+			T_{n}^{\prime}: L_{2}\left(P_{n}\right) \rightarrow L_{2}\left(P_{n}\right), T_{n}^{\prime} f(x)=\int h_{n}(x, y) f(y) d P_{n}(y) \\
+			T: L_{2}(P) \rightarrow L_{2}(P), T f(x)=\int h(x, y) f(y) d P(y)
+		\end{split}
+	\end{equation}
+	
+	The paper proves that $d_n$ converges to $d$ uniformly on the sample, and $\left\|T_n'-T_n\right\|_{L_{2}\left(P_{n}\right)}$ converges to 0. With sufficiently large sample size $n$, we can then get $d_n(x)$ arbitrarily close to $d(x)$. Thus $\left\|T_{n}-T_{n}'\right\|$ and $\left\|H_{n}-H_{n}'\right\|$ almost surely converges to 0.
+	
+	They relate the operators $T_n$ defined on $L_2 (P_n)$, to some operators $S_n$ on the space $L_2 (P)$ such that their spectra are preserved, so that it could be used as a middle ground to prove $T_n$ converges to $T$. The problem that the operators $T_n$ and $T$ are not deﬁned on the same space has been circumvented by considering bilinear forms instead of the operators themselves.
```
Comment (all2187) [inline chapter_3/2_spectral_clustering.tex]: There's a weird character here "deﬁned" (the f and i are a single character). It doesn't compile as a result. Fix this please.

File: chapter_3/2_spectral_clustering.tex [hunk 2]
```diff
@@ -128,7 +129,7 @@ \section{Spectral Clustering}
 		\item let $A \subset V$. $\mathbbm 1_A^T L \mathbbm 1_A = \sum_{(i,j) \in E} w_{ij} (\mathbbm  1_A(i - \mathbbm 1_A(j)) )^2 = weight(E(A, \bar{A})) $
 	\end{enumerate}
 	
-	\subsection{Review of Spectral Theory: Geometric \& Variational Characterization of Eigenvalues}
+	\subsection{Review of Spectral Theory: Geometric \& the collection of all Borel sets on Characterization of Eigenvalues}
```
Comment (all2187) [inline chapter_3/2_spectral_clustering.tex]: What does this mean?

File: chapter_3/2_spectral_clustering.tex [hunk 3]
```diff
@@ -153,13 +154,242 @@ \section{Spectral Clustering}
...
+		$\vdots$ \\
+		$\lambda_m = \max_{v \in S^{n-1}} Rank (v)$
 	\end{theorem}
-	Pf.
+	
+	\subsection{Spectral Clustering Algorithm}
+	
+	After introducing the graph Laplacian, we present the spectral clustering algorithm that projects data onto lower dimension space. The algorithm needs predefined number of clusters, and the similarity matrix.
+	
+	\textbf{Input:} $S$: $n\times n$ similarity matrix (on $n$ datapoints), $k$: number of clusters.
```
Comment (all2187) [inline chapter_3/2_spectral_clustering.tex]: Try to put this in a \begin{algorithm}\begin{algorithmic} environment if possible (see how its done elsewhere in the tex).

File: chapter_3/2_spectral_clustering.tex [hunk 4]
```diff
@@ -153,13 +154,242 @@ \section{Spectral Clustering}
...
+	
+	The need to calculate the eigenvector of the $n\times n$ Laplacian puts the spectral clustering algorithm at $O(n^3)$ runtime. There is an $O(n \log^3 n + m \log n)$ algorithm for Planted Partitions Model problem, as described in \cite{bsh10}.
+	
+	The Planted Partitions Model problem: for a $(p,q)$-clusetered graph $G(V,E)$, vertices are divided into disjoint clusters $C_1, \ldots, C_t$. Edges between members within clusters are included with probability $p$, and between members across clusters with probability $q$.
+	
+	Note that $C(v)$ is the cluster containing vertex $v$, and $N(G,v)$ are the neighbors of $v$.
+	
+	For $p \geq 1/2$, the following Subsquare algorithm described in \cite{bsh10} finds partitions $U_1, \ldots, U_s$ with probability $1-1/\mathrm{poly}(n)$ that for each input cluster $C_i$ of size $\Omega(\max \{qn, \log n \})$, contains a cluster $U_j$ such that $C_i=U_j$. The algorithm takes three parameters, $c_0$, $c_1$, and $\delta$.
+	
+	\begin{itemize}
```
Comment (all2187) [inline chapter_3/2_spectral_clustering.tex]: Same, try to put in an algorithm env, just makes it easier to read.

**Review by all2187 — CHANGES_REQUESTED** [review]
Good job! There are a few minor typos and changes that you should make (see comments) but otherwise it looks great. Just make the changes and post a comment here when it is done and I will merge. 

Concerning Theorems 4 and 5, I think you discuss them in sufficient detail as is for our purposes. However, I'm happy to look at them in more detail and discuss them with you if you are interested. From a quick look, it seems to me that Theorem 4 is the relatively dry technical result that the paper makes use of. It seems to be along the lines of: given arcane assumptions on a function class G, the dot product (so some similarity measure) between functions f and these G functions computed from samples converges to the true dot product. More importantly this happens unformly for all functions in G (and f?). Theorem 5 applies this to the spectral clustering case to get that the sample eigenfunction converges to the true eigenfunction. At least that's what it seems like from a quick look... Let me know if you want to discuss this further, independently from these scribes.

File: chapter_3/2_spectral_clustering.tex (hunk 2 above)
Comment (tzm41) [reply chapter_3/2_spectral_clustering.tex]: Sorry, made an accidental edit

**Comment by tzm41** [comment]
@all2187 Comments addressed.

**Review by all2187 — APPROVED** [review]
Good job! Merging now.

**Pull request merged**